### PR TITLE
Support Exclude Package with custom unsafe packages

### PR DESCRIPTION
**Pull request by hmc-cs-mdrissi**
https://github.com/jazzband/pip-tools/issues/333 Closes this issue. Custom unsafe packages allows you to exclude packages. I named the option --unsafe-packages as all logic is customizing UNSAFE_PACKAGES (motivated by this [comment](https://github.com/jazzband/pip-tools/issues/333#issuecomment-917643576)), but if people would prefer I call it exclude that works. The one difference in the two names is whether original UNSAFE_PACKAGES remain or are gone. 

As I'd ideally like setuptools/pip/etc to be in my requirements.txt file, but want to exclude other packages I prefer some way to replace the existing unsafe packages entirely. This is not a big deal to me though so if people would prefer semantic of it only adds packages to unsafe list I can do that. That is less flexible though.

I also added a flag to control whether recursive dependencies of unsafe only packages are treated as unsafe or not. The current logic for setuptools/pip is recursive dependencies only from unsafe packages are treated unsafe, but my use case for exclude is I only want to remove the exact packages I list.

Multiple occurrences of the flag are allowed. It can be used like,

pip-compile --unsafe-package foo --unsafe-package bar ...

I added two test cases to cover recursive flag being on/off.

@nvie @ssbarnea 

##### Contributor checklist

- [X] Provided the tests for the changes.
- [X] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: piptools/_compat/pip_compat.py [hunk 1]
```diff
@@ -3,18 +3,18 @@
...
 from pip._internal.req.constructors import install_req_from_parsed_requirement
 from pip._vendor.packaging.version import parse as parse_version
+from pip._vendor.requests.sessions import Session
 
 PIP_VERSION = tuple(map(int, parse_version(pip.__version__).base_version.split(".")))
 
 
 def parse_requirements(
     filename: str,
-    session: PipSession,
```
Comment (hmc-cs-mdrissi) [inline piptools/_compat/pip_compat.py:17]: Oh PipSession/Session is from pyright marking it as a type error. I can undo, but as I was working on this did want to mention pyright detects several type errors that mypy misses. I'd estimate like 1-10 type errors per file.

I think most type errors are inconsistencies in some of the pip types used. Like InstallRequirement vs InstallCandidate in find_best_match. The return type is inconsistent with actual reveal type value and while I'm guessing the two are intended to be compatible they are not currently. Here PipSession is used but the function called works with a more generic Session and one of the callers had more generic type.

**Review by atugushev — DISMISSED** [review]
Generally, it looks good. I have a concern though:

File: piptools/scripts/compile.py [hunk 2]
```diff
@@ -236,6 +236,20 @@ def _get_default_option(option_name: str) -> Any:
...
+    help="List of packages to consider unsafe. Replaces "
+    f"{', '.join(sorted(UNSAFE_PACKAGES))}. Can specify multiple times.",
+)
+@click.option(
+    "--allow-unsafe-recursive/--no-allow-unsafe-recursive",
+    is_flag=True,
+    default=False,
+    help="Determines whether dependencies that solely belong to unsafe packages "
+    "should be treated as unsafe. Default is false. This only covers direct dependencies. "
+    "Dependencies of dependencies of unsafe packages will not be marked unsafe.",
```
[the remark this answers is not in the document]
Comment (ssbarnea) [reply piptools/scripts/compile.py:251]: @atugushev Can you please remove your requested changes? I do not think that this is a good enough reason for blocking this change, especially as this behavior was already existing. As part of the review process we should try to avoid scope creeping on original PR.

IMHO, that change is fixing one problem and the other aspect is unrelated.

File: piptools/scripts/compile.py [hunk 3]
```diff
@@ -462,6 +478,8 @@ def cli(
             cache=DependencyCache(cache_dir),
             clear_caches=rebuild,
             allow_unsafe=allow_unsafe,
+            unsafe_packages=set(unsafe_packages),
```
[the remark this answers is not in the document]
Comment (ssbarnea) [reply piptools/scripts/compile.py:481]: Closing as resolved.

**Comment by di** [comment]
I see this is in the [6.5.0 milestone](https://github.com/jazzband/pip-tools/milestone/53), is this blocking a 6.5.0 release?

**Comment by ssbarnea** [comment]
I would really want to see this feature in ASAP as lack of it seriously limited tool use in several projects.

File: piptools/scripts/compile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (atugushev) [reply piptools/scripts/compile.py:251]: Removed 👍

**Comment by atugushev** [comment]
It does not block the 6.5.0. Let's release it later.

**Comment by hmc-cs-mdrissi** [comment]
Is thee anything that's recommended I should do for this PR? Should I deal with dependencies of dependencies logic or should I leave it as out of scope? If there's another commit I can do to help merge this that would be helpful. It'd be nice to use this without fork.

File: piptools/scripts/compile.py [hunk 4]
```diff
@@ -236,6 +236,20 @@ def _get_default_option(option_name: str) -> Any:
     default=True,
     help="Add options to generated file",
 )
+@click.option(
+    "--unsafe-packages",
+    multiple=True,
+    help="List of packages to consider unsafe. Replaces "
```
Comment (AndydeCleyre) [inline piptools/scripts/compile.py:242]: I think the help text here is a bit misleading regarding the required syntax, suggesting that maybe the user ought to provide a CSV. I would like to bring it more in line with the descriptions for our other options which specify `multiple=True`.

File: piptools/scripts/compile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndydeCleyre) [reply piptools/scripts/compile.py:251]: It's hard for me to be sure about details for this feature, as I haven't seen many use cases up close, but I think "only want to remove the exact packages I list" is *probably* the most useful mode here. I wouldn't be opposed to dropping the recursive flag....

But if it stays, and it only covers one level of deps, I would prefer an alternative name to "recursive," which implies more depth.

[the remark this answers is not in the document]
Comment (hmc-cs-mdrissi) [reply piptools/scripts/compile.py:251]: My personal use is I have an exact list I want to remove and there dependencies I don't intend to remove. The recursive flag was mainly added as current code partly looked at recursive dependencies. I would be happy to remove recursive flag and leave that as an out of scope feature for future if anyone has a use case for it.

File: piptools/scripts/compile.py (hunk 4 above)
Comment (hmc-cs-mdrissi) [reply piptools/scripts/compile.py]: Done. I adjusted the help to be more similar to --upgrade-package.

File: piptools/scripts/compile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmc-cs-mdrissi) [reply piptools/scripts/compile.py]: I've added a commit that drops the recursive flag. I'm happy to revert if there's a preference to keep recursive logic.

File: tests/conftest.py [hunk 5]
```diff
@@ -234,6 +234,11 @@ def pip_with_index_conf(make_pip_conf):
     )
 
 
+@pytest.fixture(autouse=True)
+def pip_with_empty_conf(make_pip_conf):
```
Comment (hmc-cs-mdrissi) [inline tests/conftest.py:238]: This was added because many unit tests make assumptions about default pip.conf. Specifically if you run unit tests on master with a pip.conf that has an extra-index-url the url will get added to output of many tests causing them to fail. This makes all tests that don't specify pip.conf file use empty one and avoids needing to modify your global pip.conf for tests.

File: tests/test_cache.py [hunk 6]
```diff
@@ -116,6 +115,3 @@ def test_reverse_dependencies(from_line, tmpdir):
         "bottom": {"middle", "top"},
         "bonus": {"top"},
     }
-
-    # Clean up our temp directory
-    rmtree(tmpdir)
```
Comment (hmc-cs-mdrissi) [inline tests/test_cache.py:121]: pytest deletes tempdirs already. The extra delete here causes one test to fail after making all tests default to use empty pip.conf. The delete here causes the other fixture to fail [here](https://github.com/jazzband/pip-tools/blob/73a8a6e4c91e2f9a90642ed46294fbe94b2cf419/tests/conftest.py#L208).

File: piptools/scripts/compile.py (hunk 4 above)
Comment (AndydeCleyre) [reply piptools/scripts/compile.py]: Thanks! Maybe the user-facing `--unsafe-packages` can be renamed as singular, `--unsafe-package`?

I'm not tied to the "unsafe" term, but either way I think it should be singular.

Comment (hmc-cs-mdrissi) [reply piptools/scripts/compile.py]: Done. I'm neutral on name and other names like --exclude-package are fine. Only specific pro to keep unsafe-package is it replaces existing unsafe package list (default excludes) of pip/setuptools.

**Comment by ssbarnea** [comment]
Any chance to resolve current conflicts? I would really want to see this feature in.

**Comment by hmc-cs-mdrissi** [comment]
I've resolved conflicts and I think tests should pass now with 2020 resolver too.

**Comment by ssbarnea** [comment]
@atugushev Please have a look at this and if ok-ish lets get it in before it needs other rebases.

File: piptools/scripts/compile.py [hunk 7]
```diff
@@ -245,6 +245,12 @@ def _get_default_option(option_name: str) -> Any:
     default=True,
     help="Add options to generated file",
 )
+@click.option(
+    "--unsafe-package",
+    multiple=True,
+    help="Specify a package to consider unsafe; may be used more than once. "
+    f"Replaces default unsafe packages, {', '.join(sorted(UNSAFE_PACKAGES))}",
```
Comment (atugushev) [inline piptools/scripts/compile.py:252]:
```suggestion
    f"Replaces default unsafe packages: {', '.join(sorted(UNSAFE_PACKAGES))}",
```

Shouldn't there be a colon?

**Review by atugushev — APPROVED** [review]
Otherwise LGTM 👍🏻

**Review by ssbarnea — APPROVED** [review]

**Comment by atugushev** [comment]
@hmc-cs-mdrissi thanks for cleaning up `reverse_dependencies()`. Could you also consider this https://github.com/hmc-cs-mdrissi/pip-tools/pull/1?

**Comment by ssbarnea** [comment]
@atugushev I would merge it as is and let him fix after. Because that the first contribution even CI is not running until one of us is approving it.

**Comment by atugushev** [comment]
> I would merge it as is and let him fix after. Because that the first contribution even CI is not running until one of us is approving it.

@ssbarnea sure! Feel free to.

**Review by atugushev — APPROVED** [review]

**Comment by ssbarnea** [comment]
@atugushev That pip cache step is hilarous almost two hours and still running https://github.com/jazzband/pip-tools/runs/7379969275?check_suite_focus=true
